### ISSUE TITLE
Unable to resolve import windows/android studio

### ISSUE THREAD

**Issue by johnnyzen** [square/wire#1325]
Hi all

I have created another ticket, as the sample project kindly supplied by @Egorand does not work for me. I still get the same issue

`unable to resolve 1 imports`

Version 3.0.2
Windows 10
Android Studio

Example project here: 

https://github.com/Egorand/wire-tastes

Many thanks

**Comment by Egorand** [comment]
Thanks! That's the link to my project though, please fix.

Would you be interested in implementing a fix? Most of us use Macs and since this is Windows-specific we won't be able to reproduce the issue.

**Comment by johnnyzen** [comment]
I included your example project as thought it would be useful for someone to try on windows... but no-one has access to a windows machine? Okies, no problem, I am happy to try and fix myself, but I am not used to working with gradle plugins.. could you help me with how to turn the project debug mode on, as I can see from the source, the debug lists the paths, which would be helpful for me to try and work out whats going.

**Comment by Egorand** [comment]
Whenever you build with `-i` or `-d` flags you should be seeing INFO or DEBUG output:
```
./gradlew generateProtos -d
```

**Comment by johnnyzen** [comment]
Excellent!! thank you so much! I will see what I can do for everyone :)

**Comment by johnnyzen** [comment]
Ok, need some help if anyone can assist:

with no sourcepath defined in build.gradle, the debugger outputs:

`wireSourceDependencies.files for dep: [C:\AndroidProjects\wire-tastes\src\main\proto]`

which seems like the default proto directory.

Which building I get:

`Cause: unable to resolve 1 imports:
  tastes/tastes.proto
searching 0 proto paths:`

BUT, when defining:

`sourcePath {
        srcDir 'src/main/proto'
    }`

The debug log shows: 

`wireSourceDependencies.files for dep: [C:\AndroidProjects\wire-tastes\src\main\proto\zen\people\person.proto, C:\AndroidProjects\wire-tastes\src\main\proto\zen\tastes\tastes.proto]`

So now its showing the the full path and filename of the proto.

But when building I get:

`Collection contains no element matching the predicate.` 

which I have no clue what this means??

So, can anyone assist with what looks right, showing and including the proto files and path in the `wireSourceDependencies.files` list or shouldnt it matter?

**Comment by johnnyzen** [comment]
Also, there was a fix a while back https://github.com/square/wire/pull/1173

This belonged in the old SchemaLoader.java, now it looks like has been refactored to https://github.com/square/wire/blob/ec7d9129215ead57fa2e3df280cb618a036fe4fe/wire-compiler/src/main/java/com/squareup/wire/schema/Root.kt

Can anyone confirm the original fix has been transferred to this refactored class? Maybe not needed, just wanted to check.

**Comment by johnnyzen** [comment]
@Egorand The bug is that windows cannot handle the path within the import statement

I can get it to work if I do:

 `import "testing\\tastes\\tastes.proto";`

but not:

`import "testing/tastes/tastes.proto";`

I narrowed the files down to either Root.kt or NewSchemaLoader.kt but thats as far as I can go in my knowledge and time.

Hope this helps to identify the problem.

**Comment by johnnyzen** [comment]
I think there is a additional issue with specifing source `sourcePath { srcDir 'src/main/proto' }` in the build gradle

**Comment by Egorand** [comment]
Perhaps we're hardcoding path separators as `/` somewhere where we should be using `File.separator`? I'm not very familiar with this part of Wire so I can't answer your questions, however the sources and Git history should have all you need. I'm also unable to repro the issue as I don't have a Windows machine.

**Comment by johnnyzen** [comment]
maybe @swankjesse can help? I think https://github.com/square/wire/blob/ec7d9129215ead57fa2e3df280cb618a036fe4fe/wire-compiler/src/main/java/com/squareup/wire/schema/Root.kt

Line 162 : `val location = Location.get(base, "${rootDirectory.relativize(descendant)}")`

should include something like `.replace(File.separator, "/")` to work on windows

**Comment by Egorand** [comment]
Does that work for your test case?

**Comment by johnnyzen** [comment]
Havent been able to test.

**Comment by johnnyzen** [comment]
@swankjesse Hi there, is this something you can help with with this issue?

**Comment by yamsergey** [comment]
Hello here. I've got the same issue with `3.2.2` on Windows. And I wonder if this line some how related to the problem [`val srcDir = (dependency.files as SourceDirectorySet).srcDirs.first {path.startsWith(it.path + "/")}`](https://github.com/square/wire/blob/545695605a95bbfde92e1a123ea972510727360c/wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WireInput.kt#L138)?

**Comment by johnnyzen** [comment]
Not had any joy on getting this resolved. @yamsergey, yes what you are saying seems correct, its def. something like that

**Comment by charleskorn** [comment]
I'm not sure if this is the same issue or a different one (if it's a different one, happy to open a separate issue). I'm having issues running on Windows when building on Linux or macOS with the exact same config works fine. It appears that something is introducing a Unix-style `/` path separator when it should be a Windows-style `\`:

```
Execution failed for task ':libs:docker-client:generateMainProtos'.
> unable to find github.com/gogo/protobuf/gogoproto/gogo.proto
    searching 0 proto paths:
      
    for file D:\a\batect\batect\libs\docker-client\build\protos/github.com\moby\buildkit\api\services\control\control.proto
  unable to find github.com/moby/buildkit/solver/pb/ops.proto
    searching 0 proto paths:
      
    for file D:\a\batect\batect\libs\docker-client\build\protos/github.com\moby\buildkit\api\services\control\control.proto
  unable to find github.com/moby/buildkit/api/types/worker.proto
    searching 0 proto paths:
      
    for file D:\a\batect\batect\libs\docker-client\build\protos/github.com\moby\buildkit\api\services\control\control.proto
  github.com\moby\buildkit\api\services\control\control.proto needs to import github.com\gogo\protobuf\gogoproto\gogo.proto
    for field sizer_all (D:\a\batect\batect\libs\docker-client\build\protos/github.com\gogo\protobuf\gogoproto\gogo.proto:69:2)
  github.com\moby\buildkit\api\services\control\control.proto needs to import github.com\gogo\protobuf\gogoproto\gogo.proto
    for field marshaler_all (D:\a\batect\batect\libs\docker-client\build\protos/github.com\gogo\protobuf\gogoproto\gogo.proto:65:2)
  github.com\moby\buildkit\api\services\control\control.proto needs to import github.com\gogo\protobuf\gogoproto\gogo.proto
    for field unmarshaler_all (D:\a\batect\batect\libs\docker-client\build\protos/github.com\gogo\protobuf\gogoproto\gogo.proto:66:2)
  github.com\moby\buildkit\api\services\control\control.proto needs to import github.com\gogo\protobuf\gogoproto\gogo.proto
    for field nullable (D:\a\batect\batect\libs\docker-client\build\protos/github.com\gogo\protobuf\gogoproto\gogo.proto:130:2)
    in field keepDuration (D:\a\batect\batect\libs\docker-client\build\protos/github.com\moby\buildkit\api\services\control\control.proto:30:2)
    in message moby.buildkit.v1.PruneRequest (D:\a\batect\batect\libs\docker-client\build\protos/github.com\moby\buildkit\api\services\control\control.proto:27:1)
...
```

`D:\a\batect\batect\libs\docker-client\build\protos` is my `sourcePath`, full configuration is in [`protobuf.gradle`](https://github.com/batect/batect/blob/af1150377d4eed784fd3f90c282ea9775831e90c/libs/docker-client/gradle/protobuf.gradle).

The full Windows output is available [in this build](https://github.com/batect/batect/runs/1240070835?check_suite_focus=true#step:6:45), and you can see the same task [passes fine](https://github.com/batect/batect/runs/1240070853?check_suite_focus=true#step:6:46) on Linux.

I'm using the latest snapshot version of the Gradle plugin (3.4.1-SNAPSHOT).

**Comment by oldergod** [comment]
Could anyone try with the current SNAPSHOT ?

**Comment by charleskorn** [comment]
> Could anyone try with the current SNAPSHOT ?

The build I linked to above is using the latest SNAPSHOT version, unless it has been updated since I posted my comment.

**Comment by oldergod** [comment]
None of us have a windows OS to take the implementation to the finish line. Maybe if we were to use a virtual OS or if someone wanted to give it a try.

**Comment by charleskorn** [comment]
Microsoft publishes development VMs at https://developer.microsoft.com/en-us/windows/downloads/virtual-machines/ - that might be a way for you to test this yourself.

**Comment by johnnyzen** [comment]
I will have another go at it. We will get this fixed!! :)

**Comment by johnnyzen** [comment]
Looks like the slashes are hardcoded here: https://github.com/square/wire/blob/master/wire-library/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Location.kt

**Comment by oldergod** [comment]
I don't think it's as simple as replacing all `/` with `File.separator` because when parsing import statements, we need to use the `/` explicitely.

**Comment by johnnyzen** [comment]
> I don't think it's as simple as replacing all `/` with `File.separator` because when parsing import statements, we need to use the `/` explicitely.

Yes, changing the slashes in Location.kt fixes one problem, but import path separator within a proto is being changed.

**Comment by bentemple** [comment]
+1 to encountering this on windows in our project as well.

**Comment by Splitframe** [comment]
> +1 to still encountering this on windows in our project as well.

Do you have a workaround other than using Linux/MacOS? I am really struggling to get this working.

**Comment by oldergod** [comment]
https://github.com/square/wire/releases/tag/4.0.0-alpha.18 should have Windows support.
I've used it successfully on a sample project but if anyone wanna give it a shot on another Windows project, please do. Let us know if it breaks so we can fix it before Wire 4.0 go stable.
